Description: libparted: uninit_bg and flex_bg features should indicate ext4
 According to tune2fs(8), the uninit_bg and flex_bg features are only
 supported by ext4, so libparted should treat their presence as indicating
 ext4.
Author: Colin Watson <cjwatson@ubuntu.com>
Bug-Ubuntu: https://bugs.launchpad.net/bugs/561599
Forwarded: http://lists.alioth.debian.org/pipermail/parted-devel/2010-April/003557.html
Last-Update: 2010-04-13

Index: b/libparted/fs/ext2/ext2_fs.h
===================================================================
--- a/libparted/fs/ext2/ext2_fs.h
+++ b/libparted/fs/ext2/ext2_fs.h
@@ -57,12 +57,14 @@
 #define EXT2_FEATURE_RO_COMPAT_SPARSE_SUPER	0x0001
 #define EXT2_FEATURE_RO_COMPAT_LARGE_FILE	0x0002
 #define EXT4_FEATURE_RO_COMPAT_HUGE_FILE	0x0008
+#define EXT4_FEATURE_RO_COMPAT_GDT_CSUM		0x0010
 #define EXT4_FEATURE_RO_COMPAT_DIR_NLINK	0x0020
 
 #define EXT2_FEATURE_INCOMPAT_FILETYPE		0x0002
 #define EXT3_FEATURE_INCOMPAT_RECOVER		0x0004
 #define EXT4_FEATURE_INCOMPAT_EXTENTS		0x0040
 #define EXT4_FEATURE_INCOMPAT_64BIT		0x0080
+#define EXT4_FEATURE_INCOMPAT_FLEX_BG		0x0200
 
 /*
  * Special inodes numbers
Index: b/libparted/fs/ext2/interface.c
===================================================================
--- a/libparted/fs/ext2/interface.c
+++ b/libparted/fs/ext2/interface.c
@@ -55,11 +55,15 @@
 		is_ext4 = ((EXT2_SUPER_FEATURE_RO_COMPAT (*sb)
 			    & EXT4_FEATURE_RO_COMPAT_HUGE_FILE)
 			   || (EXT2_SUPER_FEATURE_RO_COMPAT (*sb)
+			       & EXT4_FEATURE_RO_COMPAT_GDT_CSUM)
+			   || (EXT2_SUPER_FEATURE_RO_COMPAT (*sb)
 			       & EXT4_FEATURE_RO_COMPAT_DIR_NLINK)
 			   || (EXT2_SUPER_FEATURE_INCOMPAT (*sb)
 			       & EXT4_FEATURE_INCOMPAT_EXTENTS)
 			   || (EXT2_SUPER_FEATURE_INCOMPAT (*sb)
-			       & EXT4_FEATURE_INCOMPAT_64BIT));
+			       & EXT4_FEATURE_INCOMPAT_64BIT)
+			   || (EXT2_SUPER_FEATURE_INCOMPAT (*sb)
+			       & EXT4_FEATURE_INCOMPAT_FLEX_BG));
 		free (sb);
 
 		if (expect_ext_ver == 2 && (is_ext3 || is_ext4))
Index: b/tests/t1700-ext-probe.sh
===================================================================
--- a/tests/t1700-ext-probe.sh
+++ b/tests/t1700-ext-probe.sh
@@ -41,4 +41,23 @@
 
 done
 
+# Some features should indicate ext4 by themselves.
+for feature in uninit_bg flex_bg; do
+
+test_expect_success \
+    "create an ext3 file system" '
+    dd if=/dev/zero of=$dev bs=1024 count=4096 >/dev/null &&
+    mkfs.ext3 -F $dev >/dev/null'
+
+test_expect_success \
+    "set the $feature feature" '
+    tune2fs -O $feature $dev'
+
+test_expect_success \
+    "probe the file system, which should now be ext4" '
+    parted -s $dev print >out 2>1
+    grep -w ext4 out'
+
+done
+
 test_done
